src/pepr/operator/controllers/istio/egress-ambient.ts

```diff
@@ -2,20 +2,55 @@
  * Copyright 2025 Defense Unicorns
  * SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Defense-Unicorns-Commercial
  */
-import { V1OwnerReference } from "@kubernetes/client-node";
 import { K8s } from "pepr";
-import { Allow, IstioAuthorizationPolicy, IstioServiceEntry, K8sGateway } from "../../crd";
+import {
+  Direction,
+  IstioAuthorizationPolicy,
+  IstioServiceEntry,
+  K8sGateway,
+  RemoteGenerated,
+  UDSPackage,
+} from "../../crd";
 import { Mode } from "../../crd/generated/package-v1alpha1";
 import { purgeOrphans } from "../utils";
 import { createEgressWaypointGateway, waitForWaypointPodHealthy } from "./ambient-waypoint";
-import { generateAmbientEgressAuthorizationPolicy } from "./auth-policy";
-import { getHostPortsProtocol } from "./egress";
-import { log } from "./istio-resources";
-import { generateLocalEgressSEName, generateLocalEgressServiceEntry } from "./service-entry";
-import { HostResourceMap } from "./types";
+import { generateCentralAmbientEgressAuthorizationPolicy } from "./auth-policy";
+import { inMemoryAmbientPackageMap, remapAmbientEgressResources } from "./egress";
+import { ambientEgressNamespace, log, sharedEgressPkgId } from "./istio-resources";
+import { generateSharedAmbientServiceEntry } from "./service-entry";
 
-export const ambientEgressNamespace = "istio-egress-ambient";
-export const sharedEgressPkgId = "shared-ambient-egress-resource";
+export { ambientEgressNamespace, sharedEgressPkgId };
+
+// derive owners for a given host from the contributing package IDs
+export function deriveOwnersFromContributors(
+  host: string,
+  contributingPkgIds: string[],
+  pkgItems: UDSPackage[],
+) {
+  const ownerSaPrincipals = new Set<string>();
+  const ownerNamespaces = new Set<string>();
+  const contributing = new Set(contributingPkgIds ?? []);
+  for (const pkg of pkgItems) {
+    const ns = pkg.metadata?.namespace;
+    const name = pkg.metadata?.name;
+    if (!ns || !name) {
+      continue;
+    }
+    const id = `${name}-${ns}`;
+    const mode = pkg.spec?.network?.serviceMesh?.mode || Mode.Sidecar;
+    if (!contributing.has(id) || mode !== Mode.Ambient) continue;
+    for (const allow of pkg.spec?.network?.allow ?? []) {
+      if (allow.direction !== Direction.Egress) continue;
+      if (allow.remoteHost !== host) continue;
```

Comment:
nit: these inverted checks threw me off for a bit, feels like we could just do:

```ts
      if (allow.direction == Direction.Egress && allow.remoteHost == host) {
```

to make it more direct and explicit about what we're matching (then keeping the next if/else for serviceAccount nested in this one).